lager/lenses.hpp

```diff
@@ -123,51 +129,169 @@ auto attr(Member member)
+    return zug::comp([](auto&& f) {
+        return [&, f = LAGER_FWD(f)](auto&& whole) {
+            using T = std::decay_t<decltype(whole.value())>;
+            return f(LAGER_FWD(whole).value_or(T{}))(
+                [&](auto&& x) { return LAGER_FWD(x); });
+        };
+    });
+}
+
+namespace detail {
+
+template <
+    template <typename>
+    typename PartMeta,
+    typename Whole,
+    typename Lens,
+    typename Functor>
+decltype(auto) opt_impl(Whole&& whole, Lens&& lens, Functor&& f) {
+    using Part = typename PartMeta<std::decay_t<decltype(::lager::view(
+        std::forward<Lens>(lens),
+        std::declval<std::decay_t<decltype(whole.value())>>()))>>::type;
+
+    if (whole.has_value()) {
+        return std::invoke(
+            std::forward<Functor>(f),
+            Part{::lager::view(
+                std::forward<Lens>(lens), std::forward<Whole>(whole).value())})(
+            [&](Part part) {
+                if (part.has_value()) {
+                    return std::decay_t<Whole>{::lager::set(
+                        std::forward<Lens>(lens),
+                        std::forward<Whole>(whole).value(),
+                        std::move(part).value())};
                 } else {
-                    return std::forward<decltype(p)>(p);
+                    return std::forward<Whole>(whole);
                 }
             });
+    } else {
+        return std::invoke(std::forward<Functor>(f), Part{std::nullopt})(
+            [&](auto&&) { return std::forward<Whole>(whole); });
+    }
+}
+
+template <typename T> struct remove_opt { using type = T; };
+template <typename T> struct remove_opt<std::optional<T>> { using type = T; };
+template <typename T> using remove_opt_t = typename remove_opt<T>::type;
+
+template <typename T> struct to_opt {
+    using type = std::optional<remove_opt_t<std::decay_t<T>>>;
+};
+
+template <typename T> struct add_opt {
+    using type = std::optional<std::decay_t<T>>;
+};
+
+} // namespace detail
+
+/*!
+ * `Lens<W, P> -> Lens<[W], [P]>`
+ */
+template <typename Lens>
+auto optmap(Lens&& lens) {
+    return zug::comp([lens = std::forward<Lens>(lens)](auto&& f) {
+        return [&, f = LAGER_FWD(f)](auto&& whole) {
+            return detail::opt_impl<detail::add_opt>(LAGER_FWD(whole), lens, f);
+        };
+    });
+}
+
+/*!
+ * `Lens<W, [P]> -> Lens<[W], [P]>`
+ */
+template <typename Lens>
+auto optbind(Lens&& lens) {
+    return zug::comp([lens = std::forward<Lens>(lens)](auto&& f) {
+        return [&, f = LAGER_FWD(f)](auto&& whole) {
+            return detail::opt_impl<zug::meta::identity>(LAGER_FWD(whole), lens, f);
         };
     });
 }
 
+/*!
+ * `(Lens<W, P> | Lens<W, [P]>) -> Lens<[W], [P]>`
+ */
+template <typename Lens>
+auto optlift(Lens&& lens) {
+    return zug::comp([lens = std::forward<Lens>(lens)](auto&& f) {
+        return [&, f = LAGER_FWD(f)](auto&& whole) {
+            return detail::opt_impl<detail::to_opt>(LAGER_FWD(whole), lens, f);
+        };
+    });
+}
```

Comment:
Not sure about the naming of all above. I may prefer `map_opt`, `bind_opt`, `lift_opt`. Also, it sounds like `map_opt` and `bind_opt` are redundant, since `lift_opt` does all. If that is the only version, maybe `with_opt` is better (a bit less functional sounding, but...)

Reply:
Also if you only leave one version you can inline `opt_impl` making it all a bit cleaner.

Reply:
agreed for the naming changes. `with_opt` also does sound better.
the main problem with `with_opt` is that it may yield incorrect semantics if used on a model that contains legitimate optionals. this is the reason I'm also providing `map_opt` and `bind_opt`. it is also the reason I was thinking of deprecating `with_opt` altogether or specifying a different type, `lager::lens::focus<T>` instead of `std::optional<T>` to signal a lens that may be out of focus (aka, absent value).